sdk/python/feast/feature_store.py

```diff
@@ -1464,6 +1466,7 @@ def write_to_online_store(
                 pass
             else:
                 raise ValueError("inputs must be a dictionary or a pandas DataFrame.")
+        assert_permissions(resources=feature_view, actions=[AuthzedAction.WRITE_ONLINE])
```

Comment:
actually we requested to enforce permission authorization only in methods exposed by Feast servers, this one should not be the case

Reply:
Agreed, this needs to go in feature server's push endpoint instead of here

Reply:
There is no object fetching happening in the push [endpoint](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/feature_server.py#L116-L146).  All the logic is happening inside the [push function](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/feature_store.py#L1422-L1428) then in [write_to_online_store](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/feature_store.py#L1426C14-L1426C45) / [write_to_offline_store](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/feature_store.py#L1470C9-L1470C31) These functions are only called from the feature server.

If you want me to handle permissions in the feature server, then I am thinking of moving some functions from `feature_store` into `feature_server` or repeating some of the logic to handle it in the push endpoint. Let me know WDYT ?

Reply:
+1 for repeating logic (idk why but in this case the logic is really weird, we should probably fix it afterwards). That's what I meant when I said we'd have a performance penalty (if we implemented assertions in the servers rather than in component interfaces). We basically have two conflicting design decisions here. In order for `remote` implementations to be simple to implement, it's better if interfaces accept just names of objects rather than objects themselves. At the same time we need objects to assert permissions before calling into those interfaces. Hence, we need to acquire objects twice.

@dmartinol Maybe a solution could be introducing some sort of in-memory cache for permissions. WDYT? does that sound practical?

Reply:
You mean to cache the assertion output, not the permissions, right? Here the problem is to reduce the redundancy due to fetching missing objects, not the permissions per-se.
In any case, the pre-fetch logic is needed to populate the cache, so I would start this way and then, if we feel it's a real issue, we can add some mitigation with a caching layer that can be applied to all the different servers.

Reply:
Yeah, The first place to optimize is registry caching probably. Even now, even when it's cached, some of the lookups on cached registry still needs to go through all objects rather than do a simple dict lookup because of how cached registry is stored.

The reason I brought it up now is that if we end up deciding that we need to in fact cache assertion outputs, it might be tricky with current design when `assert_permissions` takes objects rather than names as inputs. Anyway let's cross that bridge when we come to it. btw, most probably this may only matter for feature server, so a hypothetical solution might be to introduce a simple caching layer for feature server only.